### PR TITLE
[AMDGPU][LTO] Assume closed world after linking

### PR DESCRIPTION
**Pull request by gandhi56**

### CODE REVIEW

File: llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp [hunk 1]
```diff
@@ -761,7 +761,8 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
         if (EnableLowerModuleLDS)
           PM.addPass(AMDGPULowerModuleLDSPass(*this));
         if (EnableAMDGPUAttributor && Level != OptimizationLevel::O0)
-          PM.addPass(AMDGPUAttributorPass(*this));
+          PM.addPass(
+              AMDGPUAttributorPass(*this, AMDGPUAttributorOptions{true}));
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:765]: I'd suggest to do:

```
AMDGPUAttributorOptions Options;
Options.IsClosedWorld = true;
```

And then pass the option.

**Review by shiltian — APPROVED** [review]
LGTM with one nit

File: clang/test/CodeGenCUDA/gpu-rdc-amdgpu-attrs.cu [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// RUN: clang -x hip -O3 -fgpu-rdc %s -mllvm -debug-only=amdgpu-attributor -o - | FileCheck %s
```
Comment (shiltian) [inline clang/test/CodeGenCUDA/gpu-rdc-amdgpu-attrs.cu:1]: This should not be a clang test. You can have an opt test.

**Review by shiltian — CHANGES_REQUESTED** [review]
Oops, replied the wrong PR. Sorry.

File: llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp [hunk 3]
```diff
@@ -760,8 +760,11 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
         // module is partitioned for codegen.
         if (EnableLowerModuleLDS)
           PM.addPass(AMDGPULowerModuleLDSPass(*this));
+
+        AMDGPUAttributorOptions Opts;
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:764]: Bring these into the `if` block.

File: llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp [hunk 4]
```diff
@@ -1066,6 +1066,10 @@ static bool runImpl(Module &M, AnalysisGetter &AG, TargetMachine &TM,
 
   Attributor A(Functions, InfoCache, AC);
 
+  dbgs() << "Module " << M.getName() << " is "
```
Comment (shiltian) [inline llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:1069]: Where the `LLVM_DEBUG` macro?

**Review by shiltian — APPROVED** [review]

**Comment by gandhi56** [comment]
Rebased against trunk

**Pull request merged**